negotiation/sessions/hybrid_session.py

```diff
@@ -0,0 +1,353 @@
+        action = entity_tokens[1] # first element is just start symbol
+
+        return action
+
+    '''
+    action = self.manager.choose_action(state=self.state)
+    if not action:
+        action = self.retrieve_action()
+        if not action in self.manager.available_actions(self.state):
+            action = 'unknown'
+    return action
+
+    def retrieve_action(self):
+        template = self.retrieve_response_template(None)
+        action = template['tag']
+        print 'retrieved action:', action
+        return action
+    '''
+
+    def shorten_title(self, title):
+        """If the title is too long, shorten it using a generic name for filling in the templates.
+        """
+        if len(title.split()) > 3:
+            if self.kb.category == 'car':
+                return 'car'
+            elif self.kb.category == 'housing':
+                return 'apartment'
+            elif self.kb.category == 'phone':
+                return 'phone'
+            elif self.kb.category == 'bike':
+                return 'bike'
+            else:
+                return 'item'
+        else:
+            return title
+
+    @classmethod
+    def get_fraction(cls, zero, one, fraction):
+        # see rulebased session for details
+        return one * fraction + zero * (1. - fraction)
+
+    def round_price(self, price):
+        """Round the price so that it's not too specific.
+        """
+        if price == self.target:
+            return price
+        if price > 100:
+            return int(round(price, -1))
+        if price > 1000:
+            return int(round(price, -2))
+        return int(round(price))
+
+    def estimate_bottomline(self):
+        raise NotImplementedError
+
+    def init_price(self):
+        """Initial offer.
+        """
+        raise NotImplementedError
+
+    def receive(self, event):
+        # returns an Utterance object with LF attribute
+        utterance = self.parser.parse(event, self.state)
+        # returns a list of tokens which masquerades as an utterance
+        # utterance = self.env.preprocessor.process_event(event, self.kb)
+        print 'receive:', utterance
+
+        self.state.update(self.partner, utterance)
+        intent = self.process_lf(self.agent)
+        self.lf_tokens.append(intent)
+
+        if self.bottomline is None:
+            self.bottomline = self.estimate_bottomline()
+        # self.dialogue.add_utterance(event.agent, utterance)
+
+    def process_lf(self, agent):
+        # extract the LF from the utterance attribute in the state to allow
+        # long term access to context logical forms and direct access
+        intent = state.utterance.lf['intent']  # -or- state.my_act
+        intent.append(markers.EOS)
+        # if it's a new turn
+        if len(self.lf_tokens) == 0:
+            # cat_symbol = category_to_marker[self.category_str]
+            # intent.insert(0, cat_symbol)
+            role = self.agent_to_role[agent]
```

Comment:
do we really need the agent_to_role function? it's a bit hacky... can you get it by `self.kb.role`?